pyomo/contrib/observer/model_observer.py

```diff
@@ -0,0 +1,1141 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2025
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  __________________________________________________________________________
+
+import abc
+from typing import List, Sequence, Optional
+
+from pyomo.common.config import ConfigDict, ConfigValue, document_configdict
+from pyomo.core.base.constraint import ConstraintData, Constraint
+from pyomo.core.base.sos import SOSConstraintData, SOSConstraint
+from pyomo.core.base.var import VarData
+from pyomo.core.base.param import ParamData
+from pyomo.core.base.objective import ObjectiveData, Objective
+from pyomo.core.base.block import BlockData
+from pyomo.core.base.component import ActiveComponent
+from pyomo.common.collections import ComponentMap
+from pyomo.common.timing import HierarchicalTimer
+from pyomo.contrib.solver.common.util import get_objective
+from pyomo.contrib.observer.component_collector import collect_components_from_expr
+from pyomo.common.numeric_types import native_numeric_types
+import gc
+
+
+"""
+The ModelChangeDetector is meant to be used to automatically identify changes
+in a Pyomo model or block. Here is a list of changes that will be detected. 
+Note that inactive components (e.g., constraints) are treated as "removed".
+  - new constraints that have been added to the model
+  - constraints that have been removed from the model
+  - new variables that have been detected in new or modified constraints/objectives
+  - old variables that are no longer used in any constraints/objectives
+  - new parameters that have been detected in new or modified constraints/objectives
+  - old parameters that are no longer used in any constraints/objectives
+  - new objectives that have been added to the model
+  - objectives that have been removed from the model
+  - modified constraint expressions (relies on expressions being immutable)
+  - modified objective expressions (relies on expressions being immutable)
```

Comment:
It is still weird to me that we make the client remember the mapping from Variable / Param values if they need to manually update constraints (e.g., due to fixed variables / Params changing values).  This is implicit in this documentation, too: we will update the user on *structural* changes to the expressions, but not numeric changes ... and if a numeric change causes a structural change, that is (hopefully) caught by the client (if noly they know they need to look for it)?

Can we at least make these mappings public so that the client can reuse them?

Reply:
I thought we discussed this? We can expand the API as needed to support this.

Reply:
I don't think we can make the mappings public because they should not be modified. Instead we can add methods for clients to request the information they need. I don't think that needs done as part of this PR.